ixmp/backend/base.py

```diff
@@ -0,0 +1,803 @@
+            'var').
+        """
+
+    @abstractmethod
+    def s_item_set_elements(self, s, type, name, elements):
+        """Add keys or values to item *name*.
+
+        Parameters
+        ----------
+        type : 'par' or 'set'
+        name : str
+            Name of the items.
+        elements : iterable of 4-tuple
+            The members of each tuple are:
+
+            ======= ========================== ===
+            ID      Type                       Description
+            ======= ========================== ===
+            key     str or list of str or None Set elements or value indices
+            value   float or None              Parameter value
+            unit    str or None                Unit symbol
+            comment str or None                Description of the change
+            ======= ========================== ===
+
+            If *name* is indexed by other set(s), then the number of elements
+            of each *key*, and their contents, must match the index set(s).
+            When *type* is 'set', *value* and *unit* **must** be :obj:`None`.
+
+        Raises
+        ------
+        ValueError
+            If *elements* contain invalid values, e.g. key values not in the
+            index set(s).
+        Exception
+            If the Backend encounters any error adding the elements.
+
+        See also
+        --------
+        s_init_item
+        s_item_delete_elements
+        """
+
+    @abstractmethod
+    def s_item_delete_elements(self, s, type, name, keys):
+        """Remove elements of item *name*.
+
+        Parameters
+        ----------
+        type : 'par' or 'set'
+        name : str
+        keys : iterable of iterable of str
+            If *name* is indexed by other set(s), then the number of elements
+            of each key in *keys*, and their contents, must match the index
+            set(s).
+            If *name* is a basic set, then each key must be a list containing a
+            single str, which must exist in the set.
+
+        See also
+        --------
+        s_init_item
+        s_item_set_elements
+        """
+
+    @abstractmethod
+    def s_get_meta(self, s):
+        """Return all metadata.
+
+        Returns
+        -------
+        dict (str -> any)
+            Mapping from metadata keys to values.
+
+        See also
+        --------
+        s_get_meta
+        """
+
+    @abstractmethod
+    def s_set_meta(self, s, name, value):
+        """Set a single metadata key.
+
+        Parameters
+        ----------
+        name : str
+            Metadata key name.
+        value : any
+            Value for *name*.
+
+        Returns
+        -------
+        None
+        """
+
+    @abstractmethod
+    def s_clear_solution(self, s, from_year=None):
+        """Remove data associated with a model solution.
+
+        .. todo:: Document.
+        """
+
+    # Methods for message_ix.Scenario
+
+    @abstractmethod
+    def ms_cat_list(self, ms, name):
```

Comment:
I didn't get the idea of having different methods for scenarios (`s_`-prefixed) and message scenarios (`ms_`-prefixed). It contradicts with original idea of ixmp package to be generic and potentially support many models.

Reply:
> It contradicts with original idea of ixmp package to be generic and potentially support many models.

I agree. As I say in iiasa/message_ix#249, these are *stack level violations* where both ixmp and ixmp_source are currently *not* generic, but instead tailored to message_ix.

These were always there; this PR just makes it really obvious that they should not be there.

Our goal should be eventually to remove all `ms_` methods.
- I already dealt with firstmodelyear and years_active by moving the logic into message_ix. The correspond ixmp_source methods are no longer used.
- The category/mapping set features should be moved back into ixmp.
- Backend (code like) ixmp_source should not treat ixmp.Scenario or its subclasses differently. Any special treatment of message_ix.Scenario should be **in** message_ix.Scenario.

Reply:
I see. Shouldn't then `s_` prefixes be removed? As by default methods will operate on scenario (abstract API). Isn't it clear from parameter list and/or docstrings? Maybe also mark `ms_` methods with deprecation comments as the code to be moved to `message_ix` package

Reply:
> Isn't it clear from parameter list

It should be, I guess, yes :thinking: Okay, I will make that change before merging, then.

> mark `ms_` methods with deprecation comments

I think this code might end up in ixmp.Scenario, in which case it is just a name change rather than a deprecation and removal. But if this ends up going out as-is in the upcoming release, then I will add notices to explain that.